### ISSUE TITLE
Bug la căutarea cu expresii regulate cu diacritice

### ISSUE THREAD

**Issue by CatalinFrancu** [dexonline/dexonline#58]
[ă] nu returnează niciun rezultat. Se întâmplă din cauză că motorul de expresii regulate al lui MySQL (Percona) interpretează _ă_ ca două caractere, nu ca un caracter multibiyte.

Soluția este să înlocuim, la rulare, [ăîșț] cu (ă|î|ș|ț), atunci când clasa conține diacritice. Pentru asta, trebuie adăugat și suportul pentru paranteze rotunde, care în prezent sunt ignorate.

Vezi exemplele de mai jos.

**Comment by CatalinFrancu** [comment]
În MariaDB toate expresiile de mai sus merg nativ, căci MariaDB folosește biblioteca de PCRE. În Percona (pe care îl avem în producție) nu. Există pluginuri pe care le putem instala, de exemplu [lib_mysqludf_preg](https://github.com/mysqludf/lib_mysqludf_preg), dar n-aș merge în direcția asta.

**Comment by raduborza** [comment]
Doar clasele nu merg? Că n-aș investi prea mult într-o facilitate pe care nu o folosesc mulți...

**Comment by CatalinFrancu** [comment]
Da, doar clasele. Îl las deschis și marcat ca newbie.

**Comment by LadislauStrifler** [comment]
Aș avea o rezolvare pentru asta. Mai interesează pe cineva?

**Comment by CatalinFrancu** [comment]
Sigur, de ce nu? Chiar ieri vorbeam despre căutarea [[ț]*](https://dexonline.ro/definitie/%5B%C8%9B%5D*), care returnează cuvintele care încep cu _ț_... sau cu _ș_.

**Comment by LadislauStrifler** [comment]
Ar trebui restrânsă aria de **regex** la anumite reguli simple. Exemple ca `[ț]*` se pot rezolva foarte ușor prin eliminarea parantezelor în cazul în care există o singură literă cuprinsă în opțiunea „ori” - „`[]`”. Situația se complică doar în cazul negărilor `[^...]`. Astfel, pentru soluționare aș cere o analiză, dar dacă nu există aș propune câteva orientări. În ce situație se află analiza contextului?

**Comment by CatalinFrancu** [comment]
Ei, întotdeauna poți da un exemplu mai complex -- inițial @MateiGall a semnalat problema _[țuvwxyz]*_, pe care am redus-o la _[ț]*_. Esența problemei este că motorul de expresii regulate din MySQL [nu suportă UTF-8](https://bugs.mysql.com/bug.php?id=30241) (de peste 10 ani!). La final văd o recomandare pentru [lib_mysqludf_preg](https://github.com/mysqludf/lib_mysqludf_preg), dar n-am citit nimic despre ea (folosire, eficiență etc.). Ar merita testată.

**Comment by LadislauStrifler** [comment]
Da, dar expresia `.*u(ș|ț)ă`, care nu tușește este egală cu `*u[șț]ă`, care pleznește. Deci `[țuvwxyz]*` va trebui să fie `(ț|u|v|w|x|y|z).*`  și problema ar fi rezolvată. Situația este mai complexă decât un simplu UTF-8, fiind vorba și de motorul regex implementat, negările oarecum dând o bătaie mai mare de cap. 
Eu ca o soluție, fără nicio librărie adăugată, văd reducerea expresiei la una cât mai simplă, fără negări multiple `[^șț]`. 
Dar dacă suntem ZEN, la un șablon de tipul: `[^șț]ur[șț]*`, cu un entuziasm creator s-ar putea ajunge la o interogare cu constrângeri multiple după cum urmează:
```sql 
WHERE (formNoAccent NOT LIKE 'ș%' OR formNoAccent NOT LIKE 'ț%' )
AND formNoAccent RLIKE '^(.)ur(ș|ț).*$'
```
să zicem.

**Comment by CatalinFrancu** [comment]
Dar transformările astea sună mai rău decât o bibliotecă dedicată care să implementeze suportul general pentru UTF-8. Apropo, în MariaDB merge bine-mersi. Pe server avem Percona.
